python/ray/util/worker_group.py

```diff
@@ -0,0 +1,308 @@
+import socket
+from dataclasses import dataclass
+import logging
+from typing import Callable, List, TypeVar, Optional, Dict, Type, Tuple
+
+import ray
+from ray.actor import ActorHandle
+from ray.types import ObjectRef
+
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+class BaseWorkerMixin:
+    """A class to execute arbitrary functions. Does not hold any state."""
+
+    def __execute(self, func: Callable[..., T], *args, **kwargs) -> T:
+        """Executes the input function and returns the output.
+
+        Args:
+            func (Callable): The function to execute.
+            args, kwargs: The arguments to pass into func.
+        """
+        return func(*args, **kwargs)
+
+
+@dataclass
+class WorkerMetadata:
+    """Metadata for each worker/actor.
+
+    This information is expected to stay the same throughout the lifetime of
+    actor.
+
+    Args:
+        node_id (str): ID of the node this worker is on.
+        node_ip (str): IP address of the node this worker is on.
+        hostname (str): Hostname that this worker is on.
+        gpu_ids (List[int]): List of CUDA IDs available to this worker.
+    """
+    node_id: str
+    node_ip: str
+    hostname: str
+    gpu_ids: Optional[List[int]]
+
+
+@dataclass
+class Worker:
+    """Class representing a Worker."""
+    actor: ActorHandle
+    metadata: WorkerMetadata
+
+
+def create_executable_class(executable_cls: Optional[Type] = None) -> Type:
+    """Create the executable class to use as the Ray actors."""
+    if not executable_cls:
+        return BaseWorkerMixin
+    elif issubclass(executable_cls, BaseWorkerMixin):
+        return executable_cls
+    else:
+
+        class _WrappedExecutable(executable_cls, BaseWorkerMixin):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+
+        return _WrappedExecutable
+
+
+def construct_metadata() -> WorkerMetadata:
+    """Creates metadata for this worker.
+
+    This function is expected to be run on the actor.
+    """
+    node_id = ray.get_runtime_context().node_id.hex()
+    node_ip = ray.util.get_node_ip_address()
+    hostname = socket.gethostname()
+    gpu_ids = ray.get_gpu_ids()
+
+    return WorkerMetadata(
+        node_id=node_id, node_ip=node_ip, hostname=hostname, gpu_ids=gpu_ids)
+
+
+class WorkerGroup:
+    """Group of Ray Actors that can execute arbitrary functions.
+
+    ``WorkerGroup`` launches Ray actors according to the given
```

Comment:
```
ag = ActorGroup(ActorClass)
# Broadcast
ag.increment.remote()
ray.get(ag.increment.remote())

# Single worker case
ag[0].increment.remote()

# Load balancing
# TODO: Specify in constructor or in the options? Or some other API?
ag[:].increment.remote()
ag.any().increment.remote() vs. ag.all().increment.remote()
```

Reply:
Accessing `WorkerMetadata`

```
ag = ActorGroup(ActorClass)
ag.__actor_metadata -> List[WorkerMetadata]
ag.__actor_metadata[0]
```

Reply:
The load balancing could be a separate API nested on top of this:
```python
new_handle = LoadBalancedHandle(ActorGroup(ActorClass))
ray.get(new_handle.remote())
```